### PR TITLE
sys: add uuid module for central ID/addr generation

### PR DESCRIPTION
**Pull request by haukepetersen**
Wanted to do this for a long time: this PR centralizes the generation of unique IDs/addresses. Now all network drivers use the same code to generate their hardware addresses. 

Addresses #5055 (though not all problems are solved?!):
- the issue of duplicate addresses for two devices of a kind is solved
- but having repeatable addresses depends currently on the order in which the driver init functions are called -> for this I included already means to address this (by using `uuid_custom()`), but I did not yet think of a good value to use as custom generator param for this function...

@PeterKietzmann: feel free to postpone the milestone to 2017.04...

### CODE REVIEW

**Comment by kaspar030** [comment]
255 ID's ought to be enough for everybody!

Seriously, would it be hard to use a larger type for last_used, and XOR all of it's bits?

**Comment by haukepetersen** [comment]
would be slightly more overhead in two dimensions:

1. 2 or 4 byte vs 1 byte of RAM for the `last_used` var
2. more code for the XORing, as you would need to catch `len < sizeof(last_used)`

File: sys/include/uuid.h [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_uuid
+ * @ingroup     sys
+ * @brief       Generate system wide unique IDs
```
Comment (miri64) [inline sys/include/uuid.h]: With that name it maybe woth pointing out, that these "U"UIDs are not compatible to [RFC 4122].

[RFC 4122]: https://tools.ietf.org/html/rfc4122

File: sys/include/uuid.h [hunk 2]
```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright (C) 2017 Freie Universität Berlin
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @defgroup    sys_uuid
```
Comment (miri64) [inline sys/include/uuid.h]: Name missing

Comment (haukepetersen) [inline sys/include/uuid.h]: ups

File: sys/include/uuid.h (hunk 1 above)
Comment (haukepetersen) [inline sys/include/uuid.h]: hmm, other ideas how we could name the module so there is no confusion?

Comment (miri64) [inline sys/include/uuid.h]: Maybe not applicable due to the more generalized nature of your API, but what about `devid` (or `dev_id`, so nobody will call it like some Germans call Davids ;-))

Comment (haukepetersen) [inline sys/include/uuid.h]: hm, I think a more general name would be the better fit.

File: sys/include/uuid.h [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+ *          present
+ */
+#ifndef UUID_BACKUP_SEED
+#define UUID_BACKUP_SEED        0x23
+#endif
+
+/**
+ * @brief   Get a unique ID
+ *
+ * The resulting ID is build from the base ID XORed with an 8-bit incrementing
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: `s/build from/built from`

Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: `s/base ID/base ID generated with uuid_base() which is/`

File: sys/include/uuid.h [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ * counter value into the most significant byte.
+ *
+ * @note    The resulting UUID will repeat after 256 calls.
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: -1 for the base uuid?

File: sys/include/uuid.h [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+ *
+ * @param[out] buf      memory location to copy the UUID into. MUST be able to
+ *                      hold at least @p len byte
+ * @param[in]  len      targeted length of the UUID in byte
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: `s/targeted//`
`s/byte/bytes/`

File: sys/include/uuid.h [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * @brief   Get a unique ID
+ *
+ * The resulting ID is build from the base ID XORed with an 8-bit incrementing
+ * counter value into the most significant byte.
+ *
+ * @note    The resulting UUID will repeat after 256 calls.
+ *
+ * @param[out] buf      memory location to copy the UUID into. MUST be able to
+ *                      hold at least @p len byte
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: `s/byte/bytes/`

File: sys/include/uuid.h [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+/**
+ * @brief   Get a UUID base value
+ *
+ * The uuid module creates the value dependent on the given @p len value using
+ * the cpuid module if present or a static seed value (@ref UUID_BACKUP_SEED) if
+ * not.
+ *
+ * @param[out] buf      memory location to copy the UUID into. MUST be able to
+ *                      hold at least @p len byte
+ * @param[in]  len      targeted length of the UUID in byte
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: `s/targeted//`
`s/byte/bytes/`

File: sys/include/uuid.h [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * @brief   Get a UUID base value
+ *
+ * The uuid module creates the value dependent on the given @p len value using
+ * the cpuid module if present or a static seed value (@ref UUID_BACKUP_SEED) if
+ * not.
+ *
+ * @param[out] buf      memory location to copy the UUID into. MUST be able to
+ *                      hold at least @p len byte
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: `s/byte/bytes/`

File: sys/include/uuid.h [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+ * @param[in]  len      targeted length of the UUID in byte
+ */
+void uuid_get(void *buf, size_t len);
+
+/**
+ * @brief   Get a custom unique ID based on a user given generator value
+ *
+ * The resulting ID is build from the base ID XORed with @p gen in the base
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: `s/build/built/`

File: sys/include/uuid.h [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+ * ID's most significant bytes.
+ *
+ * @note    Calling this function with identical values for @p gen and @p len
+ *          will always result in identical UUIDs.
+ *
+ * @param[out] buf      memory location to copy the UUID into. MUST be able to
+ *                      hold at least @p len byte
+ * @param[in]  len      targeted length of the UUID in byte
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: compare other comments

File: sys/include/uuid.h [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+ */
+void uuid_get(void *buf, size_t len);
+
+/**
+ * @brief   Get a custom unique ID based on a user given generator value
+ *
+ * The resulting ID is build from the base ID XORed with @p gen in the base
+ * ID's most significant bytes.
```
Comment (LudwigKnuepfer) [inline sys/include/uuid.h]: Is this really true? It's late and the code is somewhat hard to read ;)

File: sys/include/uuid.h (hunk 1 above)
Comment (haukepetersen) [inline sys/include/uuid.h]: but now that I thought about it, let's do it like that: I add that comment you requested and if we find a better name, we can refactor then.

Comment (miri64) [inline sys/include/uuid.h]: :+1:

File: sys/include/uuid.h (hunk 11 above)
Comment (haukepetersen) [inline sys/include/uuid.h]: yes.

**Comment by haukepetersen** [comment]
rebased and (hopefully) fixed doc issues

**Comment by haukepetersen** [comment]
squashed and rebased.

**Comment by haukepetersen** [comment]
rebased. Somehow Murdock failed, but can't really tell what the problem was..

**Comment by miri64** [comment]
And go

**Pull request merged**